src/java.smartcardio/share/native/libj2pcsc/pcsc.c

```diff
@@ -40,15 +40,9 @@
 // #define J2PCSC_DEBUG
 
 #ifdef J2PCSC_DEBUG
-#define dprintf(s) printf(s)
-#define dprintf1(s, p1) printf(s, p1)
-#define dprintf2(s, p1, p2) printf(s, p1, p2)
-#define dprintf3(s, p1, p2, p3) printf(s, p1, p2, p3)
+#define debug_printf(format, ...) printf(format, ##__VA_ARGS__)
```

Comment:
`##__VA_ARGS__` is a gcc extension that is also supported by clang and MSVC++
(for MSVC++, maybe only with recent versions or necessary flags that we're
using?). The standard way to accomplish this sort of thing is `__VA_OPT__`,
but that's a C++20/C23 feature, so not available to us. It seems we're already
using that extension in over 50 places in the JDK (nearly a dozon in HotSpot).
So okay.

Though I think `debug_printf(...) printf(__VA_ARGS__)` instead would work in
this case.  It's not quite as nicely self-documenting though.

Reply:
Thanks for the detailed explanation and analysis.